packages/core/src/application_init.ts

```diff
@@ -27,6 +27,80 @@ import {noop} from './util/noop';
  *
  * @see `ApplicationInitStatus`
  *
+ * @usageNotes
+ *
+ * ### Usage examples
+ *
+ * The following module provides a function returning a promise to the `APP_INITIALIZER` token.
+ *
+ * ```
+ *  function initializeApp(): Promise<any> {
+ *    return new Promise((resolve, reject) => {
+ *      // Do some asynchronous stuff
+ *      resolve();
+ *    });
+ *  }
+ *
+ *  @NgModule({
+ *   imports: [BrowserModule],
+ *   declarations: [AppComponent, HelloComponent],
+ *   bootstrap: [AppComponent],
+ *   providers: [{ provide: APP_INITIALIZER, useFactory: () => initializeApp }]
+ *   })
+ *  export class AppModule {}
+ * ```
+ *
+ * It is also possible to provide a function returning an observable to the `APP_INITIALIZER` token.
+ *
+ * ```
+ *  function initializeApp(httpClient: HttpClient): Observable<any> {
+ *   return httpClient.get("https://someUrl.com/api/user")
+ *     .pipe(
+ *        tap(user => {
+ *          localStorage.setItem("user", user);
+ *        })
+ *     )
```

Comment:
Great example of using factory fn with some deps! I'd propose to omit function logic to keep example more generic:
```suggestion
 *   return httpClient.get("https://someUrl.com/api/user")
 *     .pipe(tap(user => { /* ... */ }));
```